qutebrowser/misc/sql.py

```diff
@@ -0,0 +1,232 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016-2017 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Provides access to an in-memory sqlite database."""
+
+import collections
+
+from PyQt5.QtCore import QObject, pyqtSignal
+from PyQt5.QtSql import QSqlDatabase, QSqlQuery
+
+from qutebrowser.utils import log
+
+
+class SqlException(Exception):
+
+    """Raised on an error interacting with the SQL database."""
+
+    pass
+
+
+def init(db_path):
+    """Initialize the SQL database connection."""
+    database = QSqlDatabase.addDatabase('QSQLITE')
+    database.setDatabaseName(db_path)
+    if not database.open():
+        raise SqlException("Failed to open sqlite database at {}: {}"
+                           .format(db_path, database.lastError().text()))
+
+
+def close():
+    """Close the SQL connection."""
+    QSqlDatabase.removeDatabase(QSqlDatabase.database().connectionName())
+
+
+def version():
+    """Return the sqlite version string."""
+    q = Query("select sqlite_version()")
+    q.run()
+    return q.value()
+
+
+class Query(QSqlQuery):
+
+    """A prepared SQL Query."""
+
+    def __init__(self, querystr):
+        super().__init__(QSqlDatabase.database())
+        log.sql.debug('Preparing SQL query: "{}"'.format(querystr))
+        self.prepare(querystr)
+
+    def __iter__(self):
+        assert self.isActive(), "Cannot iterate inactive query"
+        rec = self.record()
+        fields = [rec.fieldName(i) for i in range(rec.count())]
+        rowtype = collections.namedtuple('ResultRow', fields)
+
+        while self.next():
+            rec = self.record()
+            yield rowtype(*[rec.value(i) for i in range(rec.count())])
+
+    def run(self, values=None):
```

Comment:
Looks like SQLAlchemy does actually use keyword args:

```python
>>> conn.execute(s, x='m', y='z', e1='%@aol.com', e2='%@msn.com').fetchall()
[(u'Wendy Williams, wendy@aol.com',)]
```

and only a list of dicts for batch inserts?

```python
>>> conn.execute(addresses.insert(), [
...    {'user_id': 1, 'email_address' : 'jack@yahoo.com'},
...    {'user_id': 1, 'email_address' : 'jack@msn.com'},
...    {'user_id': 2, 'email_address' : 'www@www.org'},
...    {'user_id': 2, 'email_address' : 'wendy@aol.com'},
... ])
```

that seems like a good plan - but I wonder when we end up having a quteORM 😆